keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/README.md

```diff
@@ -0,0 +1,1246 @@
+
+The linter will flag any violations of these rules, ensuring consistent zero-value handling and preventing related errors. This automated enforcement is crucial for catching issues early in the development process.
+
+<<[/UNRESOLVED]>>
+
+### Ratcheting
+
+As `validation-gen`‘s go validation code has old object access we can write any transition rule we want for validation ratcheting. `validation-gen` has access to the old and new state of obj for any field.  This gives us a building block to make any flavor of ratcheting we would need in theory.  We can think the basic form of ratcheting as - “allow old value to be written in updates even if now not valid as long as it doesn’t change -> easy to test”.  If old = new -> short circuit validation (don’t care if it failed).  If certain ratcheting needs syntactic sugar, we can add that as well based ont the current `validation-gen` design. An example of what ratcheting validation logic might look like for `validation-gen` is below:
+
+```go
+// Tightened validation (also known as ratcheting validation) would be supported by
+// defining a new validation function. For example:
+//
+        func TightenedMaxLength(opCtx operation.Context, fldPath *field.Path, value, oldValue *string) field.ErrorList {
+          if oldValue != nil && len(MaxLength(opCtx, fldPath, oldValue, nil)) > 0 {
+            // old value is not valid, so this value skips the tightened validation
+            return nil
+          }
+          return MaxLength(opCtx, fldPath, value, nil)
+        }
+
+```
+
+### Test Plan
+
+    [x] I/we understand the owners of the involved components may require updates to existing tests to make this code solid enough prior to committing the changes necessary to implement this enhancement.
+
+##### Prerequisite testing updates
+
+ None
+
+##### Unit tests
+
+###### `validation-gen` Framework And Validator Tests
+
+For the `validation-gen` code generator and framework, unit testing will be done for all core functionality and validators in `validation-gen`.  An example of the extensive units test for `validation-gen`’s validators can be seen here in the `output_tests` directory in the current prototype:
+
+https://github.com/jpbetz/kubernetes/tree/validation-gen/staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests \
+ \
+`validation-gen` validators will also use test fixtures for validator testing which allows for additional test coverage.  Example of such a generated test (`zz_generated.validations_test.go`) from the prototype is here:  \
+https://github.com/jpbetz/kubernetes/blob/validation-gen/staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/pointers/zz_generated.[validations_test.go](https://github.com/jpbetz/kubernetes/blob/validation-gen/staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/pointers/zz_generated.validations_test.go)
+
+##### Integration tests
+
+###### 	Migration Equivalency Tests
```

Comment:
Will there be a report of validation test coverage of fields in types.go.

Reply:
interesting idea.  I am not sure how to do that without writing another tool or framework.  Do you have a model in mind?

Reply:
> interesting idea. I am not sure how to do that without writing another tool or framework. Do you have a model in mind?

If it was possible to build a fuzzer based on the tags that opens a lot of possibilities.  Without that, perhaps some kind of instrumentation that wraps the validation function so we know if each returned an error at least once?

Reply:
The current POC is like API round-trip: fuzz it N times and validate equivalence.  That doesn't PROVE 100%, but we also have extensive manual tests of each tag.